extensions/src/main/java/org/pentaho/platform/osgi/KarafBoot.java

```diff
@@ -304,50 +313,57 @@ void cleanCachesIfFlagSet( String root ) throws IOException {
         }
       }
 
-      customProps.setProperty( CLEAN_KARAF_CACHE, "false" );
+      customProperties.setProperty( CLEAN_KARAF_CACHE, "false" );
+
       FileOutputStream out = null;
       try {
+        final String customLocation = root + "/etc/custom.properties";
+
         out = new FileOutputStream( customLocation );
         logger.info( "Setting " + CLEAN_KARAF_CACHE + " back to false as this is a one-time action" );
-        customProps.store( out, "Turning of one-time cache clean setting" );
+
+        customProperties.store( out, "Turning of one-time cache clean setting" );
       } finally {
+        this.karafCustomProperties = customProperties;
```

Comment:
Not really needed because `getCustomProperties()` returns the exact same reference... `customProperties.setProperty( CLEAN_KARAF_CACHE, "false" )` already changed `this.karafCustomProperties`.

Maybe you can just use `this.karafCustomProperties` everywhere, ensuring it gets instantiated at the top of the `startup()` method. We know it will always be needed (there is no code path where `/etc/custom.properties` isn't read), so that way `startup()` is responsible for setting it up and tearing it down.